pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -342,6 +345,137 @@ func setupKernelTunables(option KernelTunableBehavior) error {
 	return utilerrors.NewAggregate(errList)
 }
 
+func setQOSCgroupMemoryLimit(cgroupManager CgroupManager, qosClass v1.PodQOSClass, limit int64) error {
+	glog.V(2).Infof("[Container Manager] Setting %s QoS memory limit to %d bytes", qosClass, limit)
+	// Attempt to set specified limit
+	cgroupConfig := CgroupConfig{
+		Name: cgroupManager.CgroupName(string(qosClass)),
+		ResourceParameters: &ResourceConfig{
+			Memory: &limit,
+		},
+	}
+	err := cgroupManager.Update(&cgroupConfig)
+	if err == nil {
+		return nil
+	}
+
+	// Unreclaimable memory usage already exceeds the desired limit.
+	// Attempt to set the limit near the current usage to put pressure
+	// on the cgroup and prevent further growth.
+	glog.Warningf("[Container Manager] %s QoS memory usage is already above desired limit", qosClass)
```

Comment:
i also worry this message will be scary for a normal situation.

maybe just note that the kubelet will attempt to reach the desired state in stages and not make this a warning.